File: src/winnermenu.py
from typing import TYPE_CHECKING
from .menu import Menu
from .player import Player
from .common import PI, GameState
import pygame
import math

if TYPE_CHECKING:
    from .game import Game

class WinnerMenu(Menu):
    def __init__(self, game: 'Game'):
        super().__init__(game)
        
        self._winners = []
        self._is_draw = False
        highest_score = -1
        
        players = game.get_players()
        for i in range(game.get_num_of_players()):
            if players[i]:
                if players[i].get_score() > highest_score:
                    highest_score = players[i].get_score()
                    self._winners = [players[i]]
                    self._is_draw = False
                elif players[i].get_score() == highest_score:
                    self._is_draw = True
                    self._winners.append(players[i])
                    
        self._spinning = 0.0
        
        if game.are_human_players():
            self._time_till_active = 2.0
        else:
            self._time_till_active = 4.0

    def update(self, time: float) -> int:
        self.update_background(time)
        
        if self._time_till_active <= 0.0:
            if not self._game.are_human_players():
                self._game.delete_players()
                return GameState.MAIN_MENU
                
            players = self._game.get_players()
            for i in range(self._game.get_num_of_players()):
                if players[i] and players[i].get_command(Player.CMD_FIRE):
                    self._game.delete_players()
                    return GameState.MAIN_MENU
        else:
            self._time_till_active -= time
            
        self._spinning -= time * 4.0
        return GameState.CURRENT_STATE
        
    def draw(self):
        self.draw_background()
        title_style = self._ui.style(0.6, (1.0, 1.0, 1.0), shadow=True)
        self._ui.draw_centered_text(0.0, 6.5, "Final Result", style=title_style)

        if self._is_draw:
            self._ui.draw_centered_text(0.0, 5.5, "It's a tie!", style=title_style)
        else:
            self._ui.draw_centered_text(0.0, 5.5, "We have a winner!", style=title_style)
            
        num_winners = len(self._winners)
        
        row = 0
        col = 0
        
        cols_count = (num_winners - 1) // 4
        
        w_in_row = num_winners if num_winners < 5 else 4
        col_start = -((w_in_row - 1) * 2.0)
        
        interface = self._game.get_interface()
        
        for i in range(num_winners):
            tank = self._winners[i].get_tank()
            r, g, b = tank.get_colour()
            
            x = col_start + row * 4.0
            y = (cols_count * 2.0) - col * 4.0
            
            t_pts = [(x - 1.5, y - 0.75), (x - 0.75, y + 0.75), (x + 0.75, y + 0.75), (x + 1.5, y - 0.75)]
            self.draw_game_polygon(t_pts, (r, g, b))
            
            self._ui.draw_centered_text(x, y - 1.2, self._winners[i].get_name(), style=self._ui.style(0.4, (1.0, 1.0, 1.0), spacing=0.35, shadow=True))
            
            txt = "Winner!"
            base_angle = self._spinning
            for char_idx, char in enumerate(txt):
                angle = base_angle - (char_idx * 0.2)
                self._draw_spinning_letter(x, y - 0.4, angle, char)
            
            row += 1
            if row > 3:
                row = 0
                col += 1
                rem = num_winners - (col * 4)
                w_in_row = rem if rem < 5 else 4
                col_start = -((w_in_row - 1) * 2.0)

    def _draw_spinning_letter(self, cx, cy, angle, char):
        deg = (angle / PI) * 180.0 - 90.0
        px = cx + math.cos(angle) * 1.8
        py = cy + math.sin(angle) * 1.8
        self._ui.draw_text(px, py, char, style=self._ui.style(0.6, (1.0, 1.0, 1.0), shadow=True, proportional=False, orientation=deg))

File: groundfire-0.25/src/winnermenu.cc
////////////////////////////////////////////////////////////////////////////////
//
//               Groundfire
//
////////////////////////////////////////////////////////////////////////////////
//
// Copyright (c) 2004, Tom Russell (tom@groundfire.net)
//
// This file is part of the Groundfire project, distributed under the MIT 
// license. See the file 'COPYING', included with this distribution, for a copy
// of the full MIT licence.
//
////////////////////////////////////////////////////////////////////////////////
//
//   File name : winnermenu.cc
//
//          By : Tom Russell
//
//        Date : 28-Feb-04
//
// Description : Handles the 'We have a winner' screen
//
//
//
////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////
// Includes
////////////////////////////////////////////////////////////////////////////////
#include "winnermenu.hh"
#include "tank.hh"
#include "controls.hh"
#include "common.hh"
#include "font.hh"

#include <math.h>

////////////////////////////////////////////////////////////////////////////////
// Public Member Functions
////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////
//
// Function    : cWinnerMenu
//
// Description : Constructor
//
////////////////////////////////////////////////////////////////////////////////
cWinnerMenu::cWinnerMenu
(
    cGame * game
)
        : cMenu (game),
          _players (game->getPlayers ())
{
    int highestScore = -1;

    // Work out who won (Note: might be more than one tank)
    for (int i = 0; i < _game->getNumOfPlayers (); i++)
    {
        if (_players[i]->_score > highestScore)
        {
            highestScore = _players[i]->_score;
            _winners.clear ();
            _winners.push_back (_players[i]);
            _draw        = false;
        }
        else if (_players[i]->_score == highestScore)
        {
            _draw = true;
            _winners.push_back (_players[i]);
        }
    }

    _spinning = 0.0f;

    if (_game->areHumanPlayers ())
    {
        // Don't allow users to skip this screen until a certain time delay.
        _timeTillActive = 2.0f;
    }
    else
    {
        // No human players so wait a bit longer.
        _timeTillActive = 4.0f;
    }
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : ~cWinnerMenu
//
// Description : Destructor
//
////////////////////////////////////////////////////////////////////////////////
cWinnerMenu::~cWinnerMenu
(
)
{
    // That's all. Kill all players ready for the next game.
    _game->deletePlayers ();
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : update
//
// Description : Updates the winner menu
//
////////////////////////////////////////////////////////////////////////////////
enumGameState
cWinnerMenu::update
(
    double time
)
{
    updateBackground (time);

    if (_timeTillActive <= 0.0f)
    {
        if (!_game->areHumanPlayers ())
        {
            return (MAIN_MENU);
        }

        for (int i = 0; i < _game->getNumOfPlayers (); i++)
        {
            double when;

            if ((_game->getPlayers ())[i]->getCommand (CMD_FIRE, when))
            {
                return (MAIN_MENU);
            }
        }
    }
    else
    {
        _timeTillActive -= time;
    }

    // Update the spinning 'winner' sign.
    _spinning -= time * 4.0f;

    return (CURRENT_STATE);
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : draw
//
// Description : Draw the winner menu
//
////////////////////////////////////////////////////////////////////////////////
void
cWinnerMenu::draw
(
)
{
    drawBackground ();

    _font->setSize (0.6f, 0.6f, 0.5f);
    _font->setColour (1.0f, 1.0f, 1.0f);

    _font->setShadow (true);
    _font->printCentredAt (0.0f, 6.5f, "Final Result");
    
    if (_draw)
    {
        _font->printCentredAt (0.0f, 5.5f, "It's a tie!");
    }
    else
    {
        _font->printCentredAt (0.0f, 5.5f, "We have a winner!");
    }

    int numOfWinners = _winners.size ();

    int row = 0;
    int col = 0;

    // Work out where to position the tank graphics if we have more than one 
    // winner.
    int cols = ((numOfWinners - 1) / 4);
    float colStart = -((numOfWinners < 5 ? numOfWinners : 4) - 1) * 2.0f;

    for (int i = 0 ; i < numOfWinners; i++) 
    {
        float r, g, b;

        cTank * tank = _winners[i]->getTank();

        glLoadIdentity ();

        float x = colStart + row * 4.0f;
        float y = (cols * 2.0f) - col * 4.0f ;

        glTranslatef (x, y, -6.0f);

        tank->getColour (r, g, b);

        glColor3f (r, g, b);

        row++;
        if (row > 3)
        {
            row = 0;
            col++;
            int remainingWinners = numOfWinners - (col * 4);
            colStart = -((remainingWinners < 5 ? remainingWinners : 4) - 1) 
                * 2.0f;
        }

        glBegin (GL_QUADS);
        glVertex3f (-1.5f,  -0.75f, 0.0f);
        glVertex3f (-0.75f,  0.75f, 0.0f);
        glVertex3f ( 0.75f,  0.75f, 0.0f);
        glVertex3f ( 1.5f,  -0.75f, 0.0f);
        glEnd ();

        // Write the player's name under the tank graphic.
        _font->setSize (0.4f, 0.4f, 0.35f);
        _font->setColour (1.0f, 1.0f, 1.0f);
        _font->printCentredAt (x, y - 1.2f, "%s", _winners[i]->getName ());

        // Draw a spinning 'Winner!'
        _font->setSize (0.6f, 0.6f, 0.5f);
        _font->setColour (1.0f, 1.0f, 1.0f);
        _font->setProportional (false);

        drawSpinningLetter (x, y - 0.4f, _spinning - 0.0f, 'W');
        drawSpinningLetter (x, y - 0.4f, _spinning - 0.2f, 'i');
        drawSpinningLetter (x, y - 0.4f, _spinning - 0.4f, 'n');
        drawSpinningLetter (x, y - 0.4f, _spinning - 0.6f, 'n');
        drawSpinningLetter (x, y - 0.4f, _spinning - 0.8f, 'e');
        drawSpinningLetter (x, y - 0.4f, _spinning - 1.0f, 'r');
        drawSpinningLetter (x, y - 0.4f, _spinning - 1.2f, '!');

        _font->setOrientation (0.0f);
        _font->setProportional (true);
    }

    _font->setShadow (false);
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : drawSpinningLetter
//
// Description : Draws one letter around a tank
//
////////////////////////////////////////////////////////////////////////////////
void
cWinnerMenu::drawSpinningLetter
(
    float x,
    float y,
    float angle,
    char  charToDraw
)
{
    _font->setOrientation ((angle / PI) * 180.0f - 90.0f);
    _font->printf (x + cos (angle) * 1.8f,
                   y + sin (angle) * 1.8f,
                   "%c", charToDraw);    
}
